tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc

```diff
@@ -888,5 +889,208 @@ template <typename Combiner>
   return lhs.symbols == rhs.symbols && lhs.expr == rhs.expr;
 }
 
+SliceOpShapeHelper::SliceOpShapeHelper(Operation* op) : op(op) {
+  // mhlo.real_dynamic_slice has four operands.
+  // lmhlo.real_dynamic_slice has five oeprands, the last one is the output
+  // operand.
+  assert(op->getNumOperands() >= 4);
+  auto ty = op->getOperand(0).getType().cast<ShapedType>();
+  startIndices = SmallVector<int64_t>(ty.getRank(), kUnknown);
+  limitIndices = SmallVector<int64_t>(ty.getRank(), kUnknown);
+  strides = SmallVector<int64_t>(ty.getRank(), kUnknown);
+
+  auto attr = op->getAttrOfType<DictionaryAttr>(getAttrName());
+  if (!attr) return;
+  auto startAttr = attr.getAs<DenseIntElementsAttr>("start_indices");
+  auto limitAttr = attr.getAs<DenseIntElementsAttr>("limit_indices");
+  auto strideAttr = attr.getAs<DenseIntElementsAttr>("strides");
+  assert(startAttr.getNumElements() == ty.getRank());
+  assert(limitAttr.getNumElements() == ty.getRank());
+  assert(strideAttr.getNumElements() == ty.getRank());
+
+  for (int i = 0; i < ty.getRank(); ++i) {
+    mergeStartIndex(i, startAttr.getValues<int64_t>()[i]);
+    mergeLimitIndex(i, limitAttr.getValues<int64_t>()[i]);
+    mergeStride(i, strideAttr.getValues<int64_t>()[i]);
+  }
+}
+
+bool SliceOpShapeHelper::isFullySlicedAxis(int axis) {
+  assert(axis < static_cast<int>(startIndices.size()));
+  return startIndices[axis] == 0 && limitIndices[axis] == 1 &&
+         limitIndices[axis] == kLimitIsDimSize;
+}
+
+LogicalResult SliceOpShapeHelper::markAsFullySlicedAxis(int axis) {
+  assert(axis < static_cast<int>(startIndices.size()));
+
+  if (failed(mergeStartIndex(axis, 0)))
+    return op->emitError() << "fail to update start index for axis " << axis
+                           << "\n";
+
+  if (failed(mergeLimitIndex(axis, kLimitIsDimSize)))
+    return op->emitError() << "fail to update limit index for axis " << axis
+                           << "\n";
+
+  if (failed(mergeStride(axis, 1)))
+    return op->emitError() << "fail to update stride index for axis " << axis
+                           << "\n";
+
+  return success();
+}
+
+LogicalResult SliceOpShapeHelper::mergeStartIndex(int axis, int64_t value) {
+  assert(axis < static_cast<int>(startIndices.size()));
+
+  if (startIndices[axis] != value && value != kUnknown &&
+      startIndices[axis] != kUnknown)
+    return failure();
+
+  if (startIndices[axis] == kUnknown) startIndices[axis] = value;
```

Comment:
if `value` is `kUnknown` and `startIndices[axis]` is not, it succeeded though no merge happend. it this expected?

Reply:
fine... it's initialized to `kUnknown`.